controller/tap/server.go

```diff
@@ -469,3 +489,105 @@ func NewServer(
 
 	return s, lis, nil
 }
+
+func indexPodByIP(obj interface{}) ([]string, error) {
+	if pod, ok := obj.(*apiv1.Pod); ok {
+		return []string{pod.Status.PodIP}, nil
+	}
+	return []string{""}, fmt.Errorf("object is not a pod")
+}
+
+// hydrateIPMeta returns a map of expanded metadata labels for a given IP.
+// If no pod cound be found for that IP, it returns an empty map.
+func (s *server) hydrateIPMeta(ip *public.IPAddress) (map[string]string, error) {
+	pod, err := s.podForIP(ip)
+	if err != nil {
+		return nil, err
+	} else if pod == nil {
+		log.WithFields(log.Fields{"ip": addr.PublicIPToString(ip)}).Debugln("no pod for IP")
+		return make(map[string]string), nil
+	}
+
+	ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod)
+	return pkgK8s.GetPodLabels(ownerKind, ownerName, pod), nil
+}
+
+// podForIP returns the pod corresponding to a given IP address, if one exists.
+//
+// If multiple pods exist with the same IP address, this may be because some
+// are terminating and the IP has been assigned to a new pod. In this case, this
+// function preferrentially selects the currently running pod, if there is one;
+// otherwise, it chooses the pod that terminated the most recently (based on the
+// assumption that it's most likely to have sent a request).
+//
+// If no pods were found for the provided IP address, it returns nil. Errors are
+// returned only in the event of an error indexing the pods list.
+func (s *server) podForIP(ip *public.IPAddress) (*apiv1.Pod, error) {
```

Comment:
@kl Out of curiousity, is your concern more that inspecting the termination timestamps is too complex, or is it more about performance? I'm asking mainly because AFAICT, the tap server only has to do the extra work if it actually does encounter the edge case of multiple terminating pods and no currently-running one.

Reply:
It's the former -- that the code for finding the right pod to return is too complex for the value that it adds, and it's not worth paying the maintenance cost for that going forward.

Reply:
Okay, in that case, I'm happy to change this code to return `nil` in the case where there are multiple terminating pods.

Reply:
Alright, I've removed the termination timestamp code in 9efad8497ea418e3c559936ab31578db6570eda2. The implementation is slightly more complex than what you posted above, however, as I thought it was still worthwhile to handle pods in the `Pending` (and `Unknown`) phases correctly?

Reply:
Hmm. I don't think this function should ever return a `Pending` pod --- it can't possibly have sent the request if it hasn't yet started. 

Consider the following cases:
+ `[Pending, Terminated]`: The pending pod can't possibly have sent the request, as it hasn't started _running_ yet --- it must have come from the terminated pod before it terminated.

+ `[Pending]` (and `[Pending, Pending, Pending...]`, which AFAIK should not actually happen): Again, the pending pod can't have sent the request. Presumably, the IP either belonged to a terminated pod that has since been garbage-collected, or something outside of the cluster --- either way, we can't provide correct metadata for it, so we should return `nil`.

+ `[Terminated, Terminated...]` This is the case where we would previously try to select the most recently terminated pod. Since that is overly complex, I'm okay with removing that logic and returning `nil` in this case.

Reply:
> I opted to return nil, but it would also be fine to return the pod at index 0 or something equally simple. 

Personally, I think returning nil is more correct in this case --- I'd rather say "we can't positively identify the pod that sent the request" rather than claiming that the request was sent from a pod that we can't definitively identify.

Reply:
That breakdown of various potential pod states makes sense to me, but it's complicated enough to warrant tests for each of the cases that you enumerated.

And that said, I feel like this discussion is a bit moot. It's true that if a pod is pending, then it will not have sent any requests. Or is it? What if some of the containers came up and started making requests, but others have not, leaving the pod in a pending state? It feels like there could be dozens of confusing states for any given group of pods, but at the end of the day, if the group size is 1, we should just return it.

Reply:
@klingerf I did some research and it looks like the precise meaning of the `Pending` pod phase is that _one or more_ of the containers in the pod have yet to start. So, you're correct that a `Pending` pod may have some containers which have started and have made requests.

I think the most correct way to handle pending pods would involve testing if any of their containers have started, but that seems fairly complex. So, I'm going to go with your original suggestion of only returning  a non-running pod if `len(pods) == 1`.

Sorry if this turned into a bit of a 🚲 🏚 ...